policies/policy_service.go

```diff
@@ -328,3 +328,43 @@ func (s policiesService) ListDatasets(ctx context.Context, token string, pm Page
 	}
 	return s.repo.RetrieveAllDatasetsByOwner(ctx, ownerID, pm)
 }
+
+func (s policiesService) InactivateDatasetBySinkID(ctx context.Context, sinkID string, token string) error {
+	ownerID, err := s.identify(token)
+	if err != nil {
+		return err
+	}
+
+	if sinkID == "" {
+		return ErrMalformedEntity
+	}
+
+	datasets, err := s.repo.RetrieveAllDatasetsInternal(ctx, ownerID)
```

Comment:
You should create a method on the repository to do it based on the provided sink id. I know that you're trying to keep the single responsibility principle ok, but got through each dataset, check if it has only one sink and if the sink it's was removed, looks like a lot of steps to do the same.